src/computability/tm_to_partrec.lean

```diff
@@ -0,0 +1,885 @@
+import computability.halting
+import computability.turing_machine
+import data.num.lemmas
+
+open function (update)
+open relation
+
+namespace turing
+
+namespace to_partrec
+
+inductive code
```

Comment:
How is this related to `partrec_code`?

Reply:
It is similar in spirit, being a data type that mirrors the construction of `partrec`, but the basis here is different, using operations on `list nat` instead of `nat -> nat` as in `partrec` or `vector nat n -> nat` as in `partrec'`.

Reply:
What I actually wanted to ask is whether we should merge the two codes.  The difference in denotation (list/vector) shouldn't matter for the codes.

Reply:
Oh I see. We could push this modification of the basis up, it is true, although the proof of `exists_code` relies on `partrec'`, which is a different simplification of the basis used in `nat.partrec.code` (so there are actually three different bases here: the one on `nat ->. nat`, `vector nat n ->. nat` and `list nat ->. nat`). It's not just that the denotations are different, the basic operations show some awareness of the base type: the nat -> nat version contains functions `left, right, pair` using the cantor pairing function, the vector -> nat version uses an n-ary composition operation, and this version uses list functions like head and cons. They are all equivalent, of course, but the point of having distinct inductives is so that we can formally state this equivalence and get a usable resulting theorem.

Reply:
Ah ok, this makes sense.